src/test/java/org/sagebionetworks/web/unitclient/widget/table/v2/TotalVisibleResultsWidgetTest.java

```diff
@@ -0,0 +1,149 @@
+package org.sagebionetworks.web.unitclient.widget.table.v2;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Matchers.anyInt;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyZeroInteractions;
+import static org.sagebionetworks.repo.model.table.QueryOptions.BUNDLE_MASK_QUERY_COUNT;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+import org.sagebionetworks.repo.model.table.Dataset;
+import org.sagebionetworks.repo.model.table.DatasetItem;
+import org.sagebionetworks.repo.model.table.EntityView;
+import org.sagebionetworks.repo.model.table.QueryBundleRequest;
+import org.sagebionetworks.repo.model.table.QueryResultBundle;
+import org.sagebionetworks.web.client.PopupUtilsView;
+import org.sagebionetworks.web.client.widget.asynch.AsynchronousJobTracker;
+import org.sagebionetworks.web.client.widget.asynch.UpdatingAsynchProgressHandler;
+import org.sagebionetworks.web.client.widget.table.v2.TotalVisibleResultsWidget;
+import org.sagebionetworks.web.client.widget.table.v2.TotalVisibleResultsWidgetView;
+import org.sagebionetworks.web.shared.asynch.AsynchType;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TotalVisibleResultsWidgetTest {
+	TotalVisibleResultsWidget widget;
+
+	@Mock
+	TotalVisibleResultsWidgetView mockView;
+	@Mock
+	PopupUtilsView mockPopupUtils;
+	@Mock
+	AsynchronousJobTracker mockAsyncJobTracker;
+	@Captor
+	ArgumentCaptor<QueryBundleRequest> queryBundleRequestCaptor;
+	@Captor
+	ArgumentCaptor<UpdatingAsynchProgressHandler> asyncJobHandlerCaptor;
+
+
+	Dataset datasetLatestVersion;
+	Dataset datasetSnapshot;
+	EntityView entityView;
+
+	QueryResultBundle queryResultBundle;
+
+	List<DatasetItem> datasetItems;
+
+	private static final String TABLE_ID = "syn123";
+
+	@Before
+	public void before() {
+		widget = new TotalVisibleResultsWidget(mockView, mockPopupUtils, mockAsyncJobTracker);
+
+		datasetItems = Arrays.asList(new DatasetItem(), new DatasetItem(), new DatasetItem());
+
+		datasetLatestVersion = new Dataset();
+		datasetLatestVersion.setId(TABLE_ID);
+		datasetLatestVersion.setItems(datasetItems);
+		datasetLatestVersion.setIsLatestVersion(true);
+
+		datasetSnapshot = new Dataset();
+		datasetSnapshot.setId(TABLE_ID);
+		datasetLatestVersion.setItems(datasetItems);
+		datasetSnapshot.setIsLatestVersion(false);
+
+		entityView = new EntityView();
```

Comment:
Reasoning for concrete objects rather than mocking these (DatasetItems, Datasets, EntityView)?

Reply:
IMO it's easier to follow and debug real POJOs with data than mocks. This is also what the backend team has been doing. But if you prefer mocks, I can change these, not a big deal to me